Extended version message (xversion)

awemany opened the pull request.

awemany committed:
main: Removing never executed code regarding an old client version

A version value of 10300 is below the MIN_PEER_PROTO_VERSION,
which will generate an error already further up in the code.

awemany committed:
Extended Version Message implementation

This is a proposal and implementation for augmenting the `version`
message in the BCH protocol with an `xversion` message to provide
a key value map of 64bit integers to `std::vector<uint8_t>` values
(called `xmap`) that can be used to signal general configuration
and version options to connected peers.

This implementation comes with a specification document in
doc/xversionmessage.md, a simple unit test and a fuzzer test
case. It also comes with a proposal on how to organize and
name the `xmap` key space between implementations, as well as a tool
to process a key space definition file (xversionkeys.dat) into
a C++ header.

Furthermore, the BU-specific BUVERSION message has been retired (but
is still handled) and replaced with an appropriate entry in this map.

Thanks to everyone who gave me input on this PR:

@gandrewstone, @Greg-Griffith, @dgenr8, @sickpig, @dagurval,
@jasonbcox, @schancel, @deadalnix

gandrewstone commented:
this looks great, with one qualification.  I thought that we had talked about using a vector of bytes for the value.  I am worried that a config option, call it A, that requires more than 8 bytes will have to do something like installing multiple keys A1,A2,A3 and joining them back together.  This would be very awkward.

awemany commented:
@gandrewstone: Thank you. And, yes, correct, we have absolutely talked about a `vector<uint8_t>` for the value instead. And that was also my first implementation.

But when I actually started to use it in the place where these extended version variables make sense, the `vector<uint8_t>` was annoyingly cumbersome to deal with compared to just a 64-bit unsigned integer.  So I took the freedom to change the value type for this proposal.

I think likely 95% of version and configuration values will likely be covered with just a single `uint64_t`, don't you think?

And for the few values that need `std::vector<uint8_t>`, it should be relatively easy to hide that behind some kind of assembly scheme and accessor functions. If you want, I can add such a scheme and define it in the spec.

Of course, alternatively, one could make it `vector<uint8_t>` for the value and create a bunch of functions to reassemble `uint64_t`s from the `vector<uint8_t>`s. 

But the way it is implemented right now as just as an `std::unordered_map<uint64_t, uint64_t>`, is, I guess, about the fastest way you can look up and read values from a table on a 64-bit machine, barring a plain array. I figured it might be very common to access the configuration table from performance-critical code.

Now again, of course, you could *also* store the values that are `uint64_t` in an additional, fast-to-access `map<64,64>` or otherwise deal with them manually. But then that's extra memory and extra effort in any case. So my general idea here was to make the common case super straight forward and easy.

That all said, I can change it back to `vector<uint8_t>` if you want, of course.

gandrewstone commented:
I have no crystal ball either to determine WRT whether 64 bits will be enough.  However, if we look at the existing uses of the version message, the fields can be dropped into 2 buckets "works in 64bits" or "not":

int64:
nversion
nservices
ntime
nonce
startingheight
relayTxs

not:
addrMe  (although ipv4 could fit, these can be ipv6)
addrFrom (although ipv4 could fit, these can be ipv6)
subversion (string)

so 3/9 or 33% of the fields would need workarounds.  Given that data, I'd prefer it to be a byte array.  I think that we should be able to easily (like in 1 or 2 LOC) deserialize the byte array into whatever, because you can construct a CDataStream out of std::vector<unsigned char>.  So we can use the bitcoin deserialization code to deserialize what's in the byte vector.

To offer another option, remember that my concern is that the structure be deserializable even if some keys are not known.  So rather than using a vector of bytes, you could push a 2 byte type and then a 2 byte data length.  If the type is understood, it can be deserialized normally, if it is not understood, the client skips the next "data length" bytes.

awemany commented:
Looking at the existing `version` fields is a good point.

I'll change it back to a `std::vector` value then, with some accessor to extract uint64s (and maybe some others). Still needs some thought about maximum sizes, as this will likely allow to create a much more fragmented data structure in memory.

dgenr8 commented:
Is this meant to replace the services bits for a remote peer to condition its behavior?

For what it's worth, XT does currently check the total length of all messages and disconnects on exception.

awemany commented:
> Is this meant to replace the services bits for a remote peer to condition its behavior?

Potentially later on, yes.

> For what it's worth, XT does currently check the total length of all messages and disconnects on exception.

Hmm, that is too bad, so this isn't compatible with current XT implementations. What do you do about unknown message types?

dgenr8 commented:
Unknown types are not length checked.

awemany commented:
Are they checked in any other sense?

dgenr8 commented:
Using a new optional message solves my earlier issue and no, we have not added any extra checking of unknown message types.

Greg-Griffith commented:
@awemany Some comments and questions about your implementation.

There seems to be no tracking if a XVersionMessage was sent to a peer or not, yet a peer will keep track of if they received one and ban if they receive multiple. It seems like we might be able to unintentionally ban ourselves because of this. 

XVERACK currently sets a dos on nodes that are not marked fBUVersionSent and does not check for a XVERSION message sent. Wouldn't we want to only accept a XVERACK if we sent XVERSION and not require us to have sent a BUVersion? Since BU version is marked legacy on main.cpp 5150. To me this seems like when BUVersion is removed, older nodes wont be able to process an XVERACK message from the newer nodes.

I can not find anywhere in the proposal a definition for the behavior of receiving multiple XVERACK messages from the same peer, and this behavior isn't defined in your code. I would think we would only expect one and multiple would be considered misbehaving. i would like to note that this is also true of the BUVERACK message yet VERACK will ban for this.

Greg-Griffith commented:
Something else i just found, when NodeA receives a VERACK msg from NodeB, NodeA sends NodeB a XVERSION then a BUVERSION message. 
Assuming these are received by the peer and processed in the same order by NodeB, NodeB will set the addrFromPort for NodeA while processing the XVERSION message and then when processing the BUVERSION message it will see NodeA->addrFromPort != 0 and it will ban NodeA.

awemany commented:
@greggriffith: Good points. Would it make sense to just handle `XVERSION` and `XVERACK` without banning altogether, sending `XVERACK` as a reply to `XVERSION` but ignoring anything but the first `XVERSION` from a peer?

Can you see any DoS vectors this might open?

awemany commented:
Ok, updates:
- Rebased
- Changed max message size from 100KiB to 100KB. This should discourage "let's save bits" optimization attempts around this size limit right away.

- From Greg's review: Properly deal with `BUVERSION` and `XVERSION` now. The logic now is that the `addrFromPort` setting is set exactly once and any attempt to set it twice is logged (but ignored), in either order.  As the code is written so that `XVERSION` should be sent by a BU node before `BUVERSION`, receiving `BUVERSION` before `XVERSION` will have a separate log message noticing that this is odd. With custom clients connecting to BU, this is kept and allowed and might still happen, though.

@greggriffith: I believe the sending once and receiving once case (except for the `BUVERSION`+`XVERSION` case that you noticed) was and is properly handled but please double check.

TODO: 

- think hard whether there might be a good RPC functional test to add for this here.
- Keep track of repeated BUVERSION messages and also ban for those, to prevent any potential DoS vector (BUVERSION flood), though I think it is unlikely that this is a viable attack.

awemany committed:
test_framework: Allow to set pushbuf for higher level send_message(..)s

Allow to set the pushbuf argument also in the send_message(..)s
methods that build on top of the basic one.

awemany committed:
test_framework: Add node_regtest_dir(..) function

To allow for accessing files other than the log file in a node's
working directory as well.

awemany committed:
nodemessages: Add CompactSize de/ser

There's some duplicated code handling with the compact integer format
that can now be reworked to use this implementation. However, this commit
doesn't contain the rework yet - it just adds one more implementation ...

awemany committed:
test_framework: Add xversion and xverack message support

Add xversion / xverack messages to mininode and handle their receipt in a
simply but likely sufficient way. No changes have been made to the sending
side of the mininode in here.

awemany committed:
RPC test xversion.py for xversion handshake

Simple functional test for the new xversion handshake. Also adds this to
the default RPC test list.

Greg-Griffith commented:
@awemany looking over the latest changes, it would seem all my concerns seem to have been addressed. 

my only remaining comment is a nit, so feel free to ignore it. 
There are inconsistencies with variable tracking and naming which might get annoying later but it causes no real issue other than personal preference.

sickpig commented:
@awemany rebase needed

dagurval commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+Bitcoin Unlimited the `0x0002` prefix, BitcoinXT the `0x0003` prefix.
+Experimental or temporary features should use the `0x1000` prefix.
+
+(FIXME: List all known implementations. Feedback welcome)
```
You could just let projects claim prefixes. I'd use `0xbeef` 🥩 :smile:

awemany replied:
I was hoping to avoid collisions and clashes by doing it like I did. I can, however, go and reverse 0xbeef for fields of the DAGURVAL type, if you want? :-)

dagurval replied:
No need, I'll just pick what I want without asking permission 😛

awemany replied:
Alright, but please let us all know then whether you claimed 0xbeef (or whatever). To avoid chaos.

dagurval commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+Note that this is different from using the default encoding you would
+get when serializing through the `std::map` serializer (which is
+commonly found in `serialize.h` in most current C++ implementations based
+on the original Satoshi code).
```
Add rationale for not using the normal `std::map` serialization?

awemany replied:
Fair question. If the table is full (at its limit of 100kB) of (expected) small int values, this would save a couple hundred kB in network traffic for each node connect. I guess this is at the edge of premature optimization. If you have a strong opinion to do it the old way, I can change it.
In any case, I think the complexities of the two approaches are comparable, however.

jasonbcox replied:
Do you expect the table to become this large?  What's the order of magnitude expected for number of features/configurations for xversion to project?

awemany replied:
@jasonbcox: Good question. No I don't expect it to become this large anytime soon. I expect like a dozen parameters per "module" (e.g. Graphene) with a dozen modules or so. However, expectations might be met or they might be surpassed - and I think it is good to look ahead and be prepared. As I said, I don't see a lot of overhead to use the compact encoding. It is (mostly) already there. What was missing was a four-line loop (or so) to serialize. What I can imagine might take up data is larger, but simple tables. Does that make sense?

jasonbcox replied:
Yes, makes sense.  If the overhead is minimal, it probably doesn't matter one way or the other.

awemany replied:
Ok, so it seems that it is good as it is now. Closing this review part.

dagurval commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+A node may serialize the `xmap` it wants to announce to its peers with
+any order of the key-value pairs. When receiving an `xmap`, entries
+with duplicate keys that are later in the message should override
+values that are that are closer to the start of the message.
```
What's the use in overriding? Can't you just make duplicate keys invalid?

Greg-Griffith replied:
to clarify, you are suggesting using the first one as the valid one instead of the last? or that entire key becomes invalid and none of them get used?

awemany replied:
@dagurval: Yes, I can. Note that this mirrors the way that the default `std::map` deserialization works. So you'd have to pick one of your remarks to be more important ;-)

dagurval replied:
@awemany OK, then I agree with your choice, better to be consistent.

dagurval commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+For obvious reasons, the enumeration of entries and their expected
+types in the map should be agreed upon between implementations. For
+this reason, a separate specification that lists configuration and
+version keys is introduced, named the `xmap directory` (see below).
```
Who maintains the directory? We don't have one for the version bits, do we need one?

awemany replied:
The idea with the prefix is that there's only bidirectional communication needed between implementations for grabbing a prefix and then each implementation can decide on its own to fill in values and unilaterally merge it into a table that is kept in sync between implementations.

Version bits (and also node message bits) always create a bit of a Byzantine General's situation ...

awemany replied:
Added the other implementations from @schancel's list.

dagurval commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+
+A directory that lists the currently defined key values comes along
+with this specification and can be found in the file named
+`xversionkeys.dat`.
```
Will this spec be updated every time someone uses a new key?

awemany replied:
I was thinking until you do a release and then you send the table around to the other implementations.

dagurval commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+be just 32-bit in size and the current specification format reflects
+this. They are assumed to consist of a 16-bit prefix in the more
+significant bits of the 64-bit integer (bits 16..31) followed by a
+16-bit suffix in the least significant bits (0..15).
```
It says _at the moment_. Is there an upgrade path in the future that is backward compatible?

awemany replied:
My idea was to refrain from doing this until there's a need to do so - and rethink *at that time* whether to change anything about the process.

You can use any value in a straight-forward way *right now*. I just thought it would make sense to not do this. The longer keys might also never be needed.

dagurval commented on src/main.cpp:
```diff
@@ -4656,17 +4656,23 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
+        // The former BUVERSION message has now been integrated into the xmap field in CXVersionMessage.
+
+
         if (pfrom->nVersion >= EXPEDITED_VERSION)
```
Only BU has version >= 80002. This means `xversion` will never be sent to other clients.

awemany replied:
Good point, thank you, this goes onto my TODO list.

dagurval reviewed (commented):
This could rationalize how its better than the existing convention of expressing support protocol for extensions.

Existing examples include:
* `sendcmpct` message in BIP152
* `sendheaders` message in BIP130
* `feefilter` message in BIP133

As far as I can tell, as this is implemented now, sending `xversion` is optional at any time.

To be useful, this should only be possible during handshake. When forced to be part of the handshake, clients can quickly know what features the nodes claims to support. So can crawlers/seeders.

The handshake would be extended something like this:

- L -> R: Send version message with the local peer's version
- R -> L: Send version message back
- R -> L: Send verack message
- R:      Sets version to the minimum of the 2 versions
- L -> R: Send verack message after receiving version message from R
- L:      Sets version to the minimum of the 2 versions
- *L -> R: Send xversion*
- *R -> L: Send xversion back*
- *R -> L: Send xverack*
- *L -> R: send xverack*
(extention in italic)

You could use a existing old type `version bit` to mean "You MUST expect xversion/xverack from me".

I think the whole description of "xmap directory" is a distraction in this spec. It's not needed for the sake of protocol specification. It's especially not relevant how your directory is formatted and parsed with a script.

Maybe replace it with "Bitcoin Unlimited maintains a directory of used keys [url HERE]. It is RECOMMENDED to follow the convention in the directory when assiging a new key".

schancel commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+Bitcoin Unlimited the `0x0002` prefix, BitcoinXT the `0x0003` prefix.
+Experimental or temporary features should use the `0x1000` prefix.
+
+(FIXME: List all known implementations. Feedback welcome)
```
```
bitcore
bcash
Parity Cash
bitprim
bchd
XT
BU
ABC
```


I'm not aware of any other fully functional nodes.  bitcore doesn't mine.

awemany replied:
Thanks I'll add those.

awemany replied:
Added the missing ones in lexicographic order, keeping the existing entries intact.

schancel commented on src/netaddress.cpp:
```diff
@@ -204,18 +204,22 @@ enum Network CNetAddr::GetNetwork() const
 }
 
-std::string CNetAddr::ToStringIP() const
+std::string CNetAddr::ToStringIP(bool name_resolve) const
```
Why is this being changed with this pull request?

awemany replied:
Good catch. This stems from an earlier implementation of `xversion` (as an extension of the version message) where this behavior of ToStringIP(..) was a problem during testing. I'll take it out.

schancel commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+and version information to peers. For example, this can then be used
+to version advanced block transmission protocols like Graphene.
+
+Notably, the `version` message can not be extended as there are
```
This does not seem like a good rational to avoid extending version.  If Bitcoin XT developers don't want to change their client behavior here, then they also won't support xversion.  

If you're worried about older versions of XT on the network, that seems to be a concern about nodes which are not being maintained.  They are, in practice, their own network anyways.

dgenr8 replied:
Network clients that don't yet support this pull request (in other words, all network clients) are not "old".  XT is just an example that enforces the current protocol more completely than others insofar as it sanity checks lengths when possible.

jasonbcox replied:
An upgrade schedule can be introduced similar, but not tied, to the hardfork schedule.  Giving nodes time to update to version that support a new version message should alleviate the issue of flagging illegal version messages.

awemany replied:
@jasonbcox: Fair point, but OTOH, upgrade schedules necessitate further communication, management, scheduling and other organizational efforts. My goal here was to keep the necessary communication down to "hey guys, we made this feature and we use these keys to configure it" around. Which we can do without stepping onto each others feet.

jasonbcox replied:
I think this just shuffles the responsibility around rather than eliminating it.  Rather than putting communication pressure on the devs <-> node operator edges, introducing a new message type puts them on the dev <-> dev edges between implementations.  Considering that dev resources are spread thin, I'm not sure this is the best place to put them.

awemany replied:
Hmm, but how does your approach move comms towards devs<->node-ops? It seems to me that in either case, the dev<->node-op comm is necessary. It seems that we actually avoid dev-dev comms to an extend as it can be gradually implemented.

schancel commented on doc/xversionmessage.md:
```diff
@@ -0,0 +1,250 @@
+
+## Notes on Bitcoin Unlimited implementation details
+
+In the Bitcoin Unlimited reference implementation, the `xversion`
```
I would encourage you to focus simply on being iterable.  In practice I would implement handling this message by iterating over it as the message comes in, and recording values for the things the client understands.  Reading this into a map is probably not a good way to handle the message unless it's for some kind of lightweight statistics node.

awemany replied:
Do you mean this for the spec or the code or both? What is the advantage you foresee with that approach?

schancel replied:
Both for the spec, and for the code.  In practice the version values seem to have specific interpretation based on their key.  That means you need special parsing functions to validate the input, and then initialize any specialize logic for the specific connection.  

Given the above (and other benefits):
1. It's better to parse the values right off the bat, and ensure they are correct.
2. You can ensure that the connection is initialized properly with type-safe values.
3. You do not need to deserialize into a map object and take up that memory.

It looks like right now with your spec you could do something like this:
(Excuse me if this isn't quite right, I haven't played with the networking code yet)

```C++
size_t numRecords = ReadCompactSize(vRecv);
for(size_t i = 0; i < numRecords; i++) {
   uint64_t key = ReadCompactSize(vRecv);
   std::vector<uint8_t> value;
   vRecv >> value; // Some kind of guard on the size here?
   switch(key) { // Can use your own jump table if you feel so inclined.
      case KeyICareAbout:
           DoSpecificStuffWithValue(pfrom, value);
           break;
      default:
          ....
   }
}
```

awemany replied:
I see. The spec can be implemented to process everything your way as well, of course.

I don't see yet what you gain by doing it this way, however. As the `DoSpecificStuffWithValue(..)` will likely store the value somewhere in the particular subsystem and then there's probably some `GetValueIStoredBefore(..)` accessor that will be used within that subsystem to deal with the value in question.
Doing it like I did factors out this part.

schancel replied:
> I don't see yet what you gain by doing it this way, however. As the DoSpecificStuffWithValue(..) will likely store the value somewhere in the particular subsystem and then there's probably some GetValueIStoredBefore(..) accessor that will be used within that subsystem to deal with the value in question.

1. You avoid the overhead of allocating and parsing a map, before doing look ups into it.
2. You should be instantiating handlers for specific messages upon receiving the appropriate xversion flags.  Using an intermediate map, and keeping it around, doesn't facilitate this behavior.
3.  It facilitates ensuring all conditionals on the behavior of the map are done at the point of handling the xversion message.

> Doing it like I did factors out this part.

You need one phase where you parse into a fully allocated map, and then another phase where you do lookups for each possible thing you support.  (Or you have something that iterates over the keys like above after allocating the map).  It's both slower, and uses more memory.  From a future-proofing standpoint, keeping the map around encourages the bad behavior of doing map lookups for conditional behavior outside of the xversion message handling.  Finally, you need significantly more code to do the same thing, not limited to a specialized map object for one protocol message.

awemany replied:
> You avoid the overhead of allocating and parsing a map, before doing look ups into it.

I think the overhead is low and likely lower than ..

> You should be instantiating handlers for specific messages

... using this approach. There's some memory to keep around, yes, but code-wise it seems quite a bit simpler.

> It facilitates ensuring all conditionals on the behavior of the map are done at the point of handling the xversion message.

Yes, this is a good point. What @dagurval wrote above on proper sequencing of `xversion` (And which I am planning to do) would equally get rid of this problem, though.

That all said, I am providing this implementation here to make a case for it but am agnostic to each implementation doing it their own way, of course. Nothing should prevent you from parsing the way you want. If you find the way I am planning to do it *code-wise* unacceptable, this proposal "gracefully degrades" to hopefully keeping a table with an integer key space definition in sync between implementations plus a definition on how to send it around. Which was really my main goal here, though of course the implementation was more work.

deadalnix commented:
So, I was thinking about this. The general idea is great, however, I think we should simply use univalue for the metadata. Here is my reasoning:
1/ We don't really need this message to be very compact. It's only sent once. So optimizing binary format is not a very proposition.
2/ We'd get human readable metadata. I'm sure this can be helpful. Currently, no visualization of the data is possible unless the visualizer knows about all the tags, which defeat the purpose.
3/ Univalue is used for the RPC, so it is widely available in existing implementations.
4/ It give us a structure that can be extended all the way down. Currently every extension will have to define how to interpret the content of its binary blob, and you can bet most will not plan to do it in an extensible manner.

schancel commented:
> @deadalnix I don't think we can assume that a message containing an xmap will only be sent between peers once. In doc/xversionmessage.md awemany outlines a changeable key type which I assume would need to send a message containing an xmap to tell peers about updates resulting in multiple sends.

The spec calls that this is message is only valid if sent *once*.  It should definitely not be allowed to be changed on the fly.  Peers need to reconnect to do this and re-establish their connection with new settings.
 
> I also don't see a reason you couldn't use univalue with the way the xmap is set up right now. The values are stored in a byte vector (std::vector<uint8_t>) so by simply using univalue as the data type in the entry you can specify that the value be deserialized as univalue.

The keys are uint64, which is what @deadalnix is referring to.

Greg-Griffith commented:
> The spec calls that this is message is only valid if sent _once_. It should definitely not be allowed to be changed on the fly. Peers need to reconnect to do this and re-establish their connection with new settings.
> 
> There is good reason for this. There may be specific setup for handling different types of messages that happen when the initial connection is setup. It's a big headache to be able to change these on the fly without deallocating an entire connection -- for little to no benefit.
> 

@schancel I never said anything about XVERSION or XVERACK being sent more than once, I said that 
>I don't think we can assume that ***a message containing an xmap*** will only be sent between peers once.

This is referring to any potential message that may use xmap. The spec mentions a second key type called 'changeable' that has the potential to be updated throughout the lifetime of the connection. I acknowledge that @awemany has clearly said they are TBD but I do not think that at this point we should operate under the assumption that they will never be implemented/used. To me that just seems like throwing away potential useful utility. To give an example, there potential additional messages, lets call them XUPDATE, that would be sent between peers to update the suggested changeable keys instead of the permanent ones. Since the initial set of keys was sent using an xmap, I don't see a reason why we wouldn't also send an xmap to update changeable keys. 

> The keys are uint64, which is what @deadalnix is referring to.

@schancel the keys are not required to be interpreted as uint64 and since @deadalnix specifically mentioned interpretation of the binary blobs 
> Currently every extension will have to define how to interpret the content of its binary blob

I don't think he thought the keys need to be uint64 either. 



Some comments for @awemany 
After re-reading the spec, the wording of the first sentence of the *Encoding of the xversion and xverack message types* implies that keys can only be uint64. I am assuming this wording was simply something that was missed when zerg asked about the vector implementation in the first 5 comments in this thread. Can you please clean this up?

Additionally, the part about what values are valid for a key's data type (*xmap directory format* section) make it seem like it they are restricted to only ever using the uint64 or vector types. I feel that this should probably cleaned up as well.

awemany commented:
> @awemany, I don't see any significant criticism. Comments such as:
> 
>     1. deserialization that ignores unknown fields, and
> 
>     2. UniValue values
> 
> 
> seem compatible with our proposal. Can we get this rebased soon in preparation for a merge?

Absolutely. There's a couple doable TODOs from above which I'll implement as well before pushing a rebased version here. Working on it now.

awemany commented:
@schancel:

> The spec calls that this is message is only valid if sent once. It should definitely not be allowed to be changed on the fly. Peers need to reconnect to do this and re-establish their connection with new settings.

> There is good reason for this. There may be specific setup for handling different types of messages that happen when the initial connection is setup. It's a big headache to be able to change these on the fly without deallocating an entire connection -- for little to no benefit.

Agreed, and what @Greg-Griffith said. But the idea is to just mark keys as changeable explicitly. I can well imagine that is never used. It is also not supported yet. Think of the map rather as a high level communication document that simply *explicitly* states individually that a value is per session. The 'changeable' part does not need to be supported (yet, and maybe never).
To address the concern in general, `xversion` is going to be properly sequenced to be at the beginning of a communication exchange (which is still TODO).

> The keys are uint64, which is what @deadalnix is referring to.

Hmm, ok so @deadalnix or you are suggesting something different to `uint64_t` **keys**? I think that would create a mess, actually. I was thinking this is about the `value` type - and I am fine, of course, if that becomes JSON for some values down the road.

awemany committed:
Rework of the initial connection handshake

This is a rather comprehensive rework of the initial connection
set up logic. Instead of having various flags that track the individual
sending and receiving of messages, the set up sequencing for a connection
has been replaced with two state machines and corresponding state
variables (state_incoming, state_outgoing) that track the status of
the incoming and outgoing connection setup parts, and of sending
any kind of version message (version, buversion, xversion) around.

This change comes with a few behavioral changes and thus needs to be
tested comprehensively in a test net as well as live nodes for reasonable
certainty as to bug scarcity.

The incoming and outgoing state variables and their allowed
transitions are as follows:

For INCOMING versioning message from the remote peer, using
the ConnectionStateIncoming enum as defined in net.h:

CONNECTED_WAIT_VERSION: Initial state. The TCP connection is up and
the node is waiting for the version message from the remote peer. Only
a version message is valid here. All other messages are invalid and
will either lead to disconnection and banning or are silently ignored.

SENT_VERACK_READY_FOR_XVERSION: After receiving the version message (and
sending the verack out in the event handler). Only now it is permissible
to receive and extended version message 'xversion' from the remote peer.
If any other message comes in, the state transitions to READY automatically,
which will make receipt of xversion messages invalid.

READY: The version handshake is complete and normal message exchange can
commence.

For OUTGOING versioning messages to the remote peer, using the
ConnectionStateOutgoing enum as defined in net.h:

CONNECTED: Initial state after TCP connection is up.

SENT_VERSION: The initial version message has been sent to the remote peer.

READY: The verack has been received and the buversion as well as xversion
messages have been sent to the remote peer.

Any regular message that is coming in the initial version message, before both
incoming and outgoing connection state transitioned to "READY" will be dropped now.

awemany committed:
toString(..) function for printing/logging of enums

This function will take an enum (cast to uint64_t) and a map
of bit values to strings and produce a string describing the OR-ed
enum value.

To be used in the following xversion commits.

awemany committed:
Disable parallel processing for extended version handshake

Only enable parallel message processing when all version information has
been exchanged.

awemany committed:
test_framework/mininode: More info on message handling errors

awemany committed:
sendheaders: Fixes for xversion

Wait for proper xversion setup before testing the sendheaders functionality.

awemany commented:
Ok, some updates:

- I have largely reworked the initial connection handshake. This is a pretty large, additional change set that might have introduced nasty bugs, so please review carefully. It now uses two state machines, one for 'being informed about the other node's version information' and one for 'informing the other side about this node's version information'. A node is deemed in `successfullyConnected()` state now only if both state machines are in their final state. Please also read the corresponding commit message.
- This rework of the initial handshake also makes `xversion` receipt mandatory to come just after `verack` (if it comes; it is still optional), anything else is deemed bannable offense, and should address @dagurval's and @schancel's concern here. 

- As a side note: It might be a good idea to allow an initial `xversion` version exchange, dropping exchange of a `version` message altogether and phasing it out eventually. This would need some cross-implementation coordination, however.

- As a further side note, the `filterxthin` and the `sendheaders` initial configuration variable exchange should also be integrated into `xversion` to be able to further reduce the amount of code later.

- The `sendheaders.py` RPC test had to be updated to properly wait for `xversion` to make its logic work again (and I hope this change is still testing all the things that makes a node compatible with other nodes on the network - but much more live testing is required on this front).

- Support for `xversion`/`xverack` has been added to the mininode.

awemany committed:
[minor] test_framework: Remove superfluous 'pass' statement

awemany committed:
test_framework: Fix repr() of bu_version class

awemany committed:
test_framework: Make sending of initial version message optional

For testing this part of the p2p handshake.

awemany commented:
Now added a simple test of the initial `xversion` handshake. 

TODO: Check new handshake to be compatible with current BU, ABC and XT implementations.

awemany commented:
Ok, further update, I did a very simple cross-implementation test between ABC and BU and the nodes do not seem to want ban each other and the ABC<->BU comms gracefully degrade to no `xversion` support, as expected.

This is the setup:
Start ABC node like this:
` ./bitcoind -regtest -rpcport=10000 -connect=localhost -debug -datadir=/tmp/abc-node`

Start BU node with `xversion` support like this:
`./bitcoind -regtest -debug -rpcport=10001 -listen -datadir=/tmp/bu-node`

And then send some blocks in either direction and do a `grep -in ban <debug.logs>` to check whether any banning messages did appear. The ABC node notices the unknown `xversion` message, as expected.
 
Also tested vice versa with respect to the `-listen` and `-connect` options.

gandrewstone commented:
taking this out of WIP

awemany commented:
Ok, I think this ready for final review. Did another two of the above test pairs with BU current `dev` as well as with XT and they seem to talk to each other just fine. Pushed version contains the new, extended list of clients (as per @schancel above).

awemany committed:
Makefile: Include xversion infrastructure as EXTRA_DIST

awemany commented:
Added the necessary files to the `Makefile.am`'s `EXTRA_DIST` as per PM from @gandrewstone. Have checked that they appear in the `.tar.gz` so this hopefully re-enables `gitian` builds.

gandrewstone commented on src/main.cpp:
```diff
@@ -1145,6 +1186,7 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
                 pfrom->fDisconnect = true;
             }
         }
+        pfrom->state_incoming = ConnectionStateIncoming::SENT_VERACK_READY_FOR_POTENTIAL_XVERSION;
```
This must be set before "pfrom->PushMessage(NetMsgType::VERACK);" or we could legitimately receive the XVERSION response before we set this state flag due to the multithreaded network processing.

gandrewstone commented on src/main.cpp:
```diff
@@ -1014,22 +1086,31 @@ bool ProcessMessage(CNode *pfrom, std::string strCommand, CDataStream &vRecv, in
+     if anything but an xversion comes in whilst in
+     SENT_VERACK_READY_FOR_POTENTIAL_XVERSION, transition to READY state and
+     thus disallow any further incoming xversions. */
+    if (pfrom->state_incoming == ConnectionStateIncoming::SENT_VERACK_READY_FOR_POTENTIAL_XVERSION &&
```
If you look at the VERACK handler, XVERSION, BUVERSION, and FILTERSIZEXTHIN are all sent.  Due to the multi-threaded processing, these could be processed in any order. Therefore, you need make this if statement tolerant of BUVERSION and FILTERSIZEXTHIN messages.  I think that doing so still constrains XVERSION to the beginning of the protocol, so its ok.

gandrewstone commented on src/net.cpp:
```diff
@@ -2144,31 +2198,34 @@ void ThreadMessageHandler()
+            else
+            {
+                // serial processing during setup
+                TRY_LOCK(pnode->csSerialPhase, lockSerial);
```
ok I see that you have serialized processing during setup (and I like the idea because it makes things easier to debug), so my prior 2 comments are not strictly necessary.  However, there seems to be no reason not to have them, especially setting state_incoming before sending VERACK.

gandrewstone commented:
My 2 small comments can be addressed in a subsequent PR, since initial message protocol is serialized here, merging.

zander commented:
I was asked to look at this.

I can't find anywhere that it explains why this is superior to just adding new message-types.

The specific reason this is relevant is future extensibility. Adding new messages is trivial and useful. It is known to work and correctly functioning clients ignore messages they don't understand. Versioning is a solved problem.

Versioning in the 'xversion' setup seems an unsolved problem and it you can't add new fields to an existing message-type (well, with LOTS of complexity you may be able to solve it again).

What is the problem that is **solved** in this ?
Solved as opposed to moved.

Greg-Griffith commented:
@zander i suppose your message is in response to my reddit post.

XVersion is an improvement on the BUVersion message, which has been deprecated. The xversion message transports a generic key-value map that is meant to hold the configuration and version parameters. The current version/verack handshake requires a HF in the event of an update, xversion solves that. It is forward compatible with any possible additional initial connection information we might want to share with peers by simply adding an entry into the xversion map.

zander commented:
>  The current version/verack handshake requires a HF in the event of an update, xversion solves that.

No, you just add a new message which is completely forward and backwards compatible.

>  It is forward compatible with any possible additional initial connection information we might want to share with peers by simply adding an entry into the xversion map.

How is this an improvement over adding a new message-type?

Your spec states that you enable new data (not map-data) to be added at the end of the message, this is not forward compatible. Old clients will end up chocking on the data you send that they don't understand.

I have not seen any problems described that this solves. Just moving it from message-types to map-keys. Adding complexity but not solving problems.

Greg-Griffith commented:
> No, you just add a new message which is completely forward and backwards compatible.

Yes, but this in turn removes the need for the version message to ever be updated because any changes that you would make to it could just be added in this extension message. Or you could do a one time HF to replace version with xversion but for now this is a no HF solution. 

> How is this an improvement over adding a new message-type?

Imo adding an entry to the map is simpler than adding a new message that needs to be exchanged for every protocol that may need it. 

> Your spec states that you enable new data (not map-data) to be added at the end of the message, this is not forward compatible. Old clients will end up chocking on the data you send that they don't understand.

They shouldn't if they have implemented the xmap correctly. if it find that it cannot read the data it silently fails and returns a default value. It should be completely forward compatible with anything short of changing prefix-suffix sizes, but i highly doubt it will ever reach that point. 

> I have not seen any problems described that this solves. Just moving it from message-types to map-keys. Adding complexity but not solving problems.
> If there are problems this solves, please share them!

You might not see this as an issue, but imo the current service bits system that we use to identify protocols is of poor design for a multiple implementation network. There must be coordination between all clients on what bit is used for what protocol before that protocol hits the network. Xversion lets each client use its own prefix to avoid collisions while still letting any client accept a protocol from another if they want to add it. 
Example: only BU has graphene right now, and it has multiple versions due to it being updated over time and would require multiple service bits (due to multiple versions that treat some messages differently) that everyone would need to coordinate on in order to be implemented by the general network. xversion lets us add/remove map keys being used for graphene as needed without needing to coordinate with everyone else because it is on the BU suffix. if another implementation wanted to also add support for graphene and they were using xversion they would only need to start reading and broadcasting the correct key to "sign up" for graphene messages from peers on the network who also support graphene. 
in short, it gives each implementation more freedom with additional protocols because they have their own prefix without needing to coordinate with every other implementation to try to reserve service bits.

gandrewstone commented:
xversion itself showed that people aren't necessarily implementing "ignore-unknown-messages".  The xversion spec explicitly specifies this behavior (and also doing so is more "natural" to people as a key/value pair rather than a message) so it is much more likely to be followed.

And the CB/Xthin service bit conflict shows how poorly it works in the context where dev groups aren't in 100% agreement.

zander commented:
Thanks for the replies, I have still not seen any problems you solved, just moved some around. Adding xversion seems to me a net negative for the protocol.

But there are some issues that you bring up that seem to be based on bad decisions in the past, and I feel I need to reply to that.

> >  No, you just add a new message which is completely forward and backwards compatible.

> Yes, but this in turn removes the need for the version message to ever be updated because any changes that you would make to it could just be added in this extension message.

You are wrong, that would not be backwards compatible. Old clients would stop getting the information they need.

> Imo adding an entry to the map is simpler than adding a new message that needs to be exchanged for every protocol that may need it.

I understand that it is your opinion, but can you explain why its easier? I don't see it.

Did xmessage code add a framework to make it easier? Are you unfamiliar with adding new message-types in the current codebase? And if it so, can't it be made equally easy?

> You might not see this as an issue, but imo the current service bits system that we use to identify protocols is of poor design for a multiple implementation network. There must be coordination between all clients on what bit is used for what protocol before that protocol hits the network. Xversion lets each client use its own prefix to avoid collisions while still letting any client accept a protocol from another if they want to add it.

Service bits are a handshake concept that were initially designed to allow full nodes to communicate with non-full nodes. I agree its not applicable if you want to use one bit per version of a certain block-relay system. Don't do that.

It is also a developer (designer?) problem that both Core and BU ended up using the same bit. I mean, you have dozens of bits free, why on earth are you taking the next one in the enum? Just roll a couple of dice and take any bit. One in 60 chance, and you choose the most predictable one...  Blaming that on the software layer is maybe not being honest. The same issue and the same solution is applicable everywhere. An INV-type enum value, for instance. Just use a random number, no need to be sequential.

xversion sounds like a very complex solution to problems that can be solved without any of the overhead you introduced.

* You have 64 service bits, less than 10 assigned. Coordination isn't really needed, just communicate well after you chose.
* should your block-relay need a handshake, then create a message (and maybe a reply) to do so.
* prefix your message-ID with your client, or just pick them smartly in general. Its a string, naming conflicts are easy to avoid.

Adding a message-type is really quite trivial, and if its too hard on the code level I suggest abstracting it into a class and making it easier in the code. Honestly, this is the parsing of the VERSION message;

```
        int64_t nTime;
        CAddress addrMe;
        CAddress addrFrom;
        uint64_t nNonce = 1;
        vRecv >> pfrom->nVersion >> pfrom->nServices >> nTime >> addrMe;
```

Nothing more. Writing this message is about the same complexity. Claiming this is hard makes no sense to me.

Your graphene example is trivial to solve using a new message type "graphene-capabilities"
With a "minimum version" and a "maximum supported version" int-pair and maybe other booleans that your algorithms can use.  
You only send this if the service bit was set, there is generally no reply needed as the other side does the same in reply to the VERSION message.
Probably takes about half an hour to implement.

If you think, Oh, I have to update protocol.c/h and main.cpp and maybe other files just to support one new message, then I think you should consider refactoring this stuff and lower the complexity on the BU codebase instead of increasing the complexity for everyone by introducing a new layer of abstraction on the network.

It should also be mentioned that to me it looks like BU isn't releasing more than one Graphene version per 6 months to the wider public. Which means that you have no requirement for heavy versioning in the first place since there are hard forks every 6 months. So this problem is one of planning when to do this upgrade, not any real versioning problem.

Hope that helps.

gandrewstone commented:
Sending a new message type will get you banned by some clients.  Adding new
fields to the version message will get you banned by some clients.

WRT service bits, even if random bits are chosen, there is still a chance
of collision which increases dramatically as new features are added.  The
number of service bits are limited so that limits the number of new
services.

We expect XVERSION to communicate more than single "on/off" service bits.
It can also include settings/configuration for different services.  For
example, a node could communicate relay policy so clients can figure out
appropriate fees upon connection.  Or communicate a succinct description of
what historical blocks a pruned node contains.  Yes we could have an
individual and separate message for each of these.  But the complexity is
greater than just serialization and deser.  What if the node doesn't send
you the message?  What if the node sends the message in the middle of the
connection lifecycle?  The XVERSION message handles these issues for you,
and its not that complex.

As the reference, we implemented a very convenient, full-features xversion
with a definition file that generates a .h.  But you don't have to do
this.  You could just deser xversion into a map and pull out whatever
hard-coded keys you are interested in.

You must be aware that we have added new messages to the protocol since
XVERSION itself is one.  So why are you telling us how to add new
messages?  This weirdness is making me wonder if you are arguing for the
sake of it.  Regardless, XVERSION has been in BU since the last release so
this ship has sailed.



On Mon, Apr 22, 2019 at 6:04 PM Tom <notifications@github.com> wrote:

> Thanks for the replies, I have still not seen any problems you solved,
> just moved some around. Adding xversion seems to me a net negative for the
> protocol.
>
> But there are some issues that you bring up that seem to be based on bad
> decisions in the past, and I feel I need to reply to that.
>
> No, you just add a new message which is completely forward and backwards
> compatible.
>
> Yes, but this in turn removes the need for the version message to ever be
> updated because any changes that you would make to it could just be added
> in this extension message.
>
> You are wrong, that would not be backwards compatible. Old clients would
> stop getting the information they need.
>
> Imo adding an entry to the map is simpler than adding a new message that
> needs to be exchanged for every protocol that may need it.
>
> I understand that it is your opinion, but can you explain why its easier?
> I don't see it.
>
> Did xmessage code add a framework to make it easier? Are you unfamiliar
> with adding new message-types in the current codebase? And if it so, can't
> it be made equally easy?
>
> You might not see this as an issue, but imo the current service bits
> system that we use to identify protocols is of poor design for a multiple
> implementation network. There must be coordination between all clients on
> what bit is used for what protocol before that protocol hits the network.
> Xversion lets each client use its own prefix to avoid collisions while
> still letting any client accept a protocol from another if they want to add
> it.
>
> Service bits are a handshake concept that were initially designed to allow
> full nodes to communicate with non-full nodes. I agree its not applicable
> if you want to use one bit per version of a certain block-relay system.
> Don't do that.
>
> It is also a developer (designer?) problem that both Core and BU ended up
> using the same bit. I mean, you have dozens of bits free, why on earth are
> you taking the next one in the enum? Just roll a couple of dice and take
> any bit. One in 60 chance, and you choose the most predictable one...
> Blaming that on the software layer is maybe not being honest. The same
> issue and the same solution is applicable everywhere. An INV-type enum
> value, for instance. Just use a random number, no need to be sequential.
>
> xversion sounds like a very complex solution to problems that can be
> solved without any of the overhead you introduced.
>
>    - You have 64 service bits, less than 10 assigned. Coordination isn't
>    really needed, just communicate well after you chose.
>    - should your block-relay need a handshake, then create a message (and
>    maybe a reply) to do so.
>    - prefix your message-ID with your client, or just pick them smartly
>    in general. Its a string, naming conflicts are easy to avoid.
>
> Adding a message-type is really quite trivial, and if its too hard on the
> code level I suggest abstracting it into a class and making it easier in
> the code. Honestly, this is the parsing of the VERSION message;
>
>         int64_t nTime;
>         CAddress addrMe;
>         CAddress addrFrom;
>         uint64_t nNonce = 1;
>         vRecv >> pfrom->nVersion >> pfrom->nServices >> nTime >> addrMe;
>
> Nothing more. Writing this message is about the same complexity. Claiming
> this is hard makes no sense to me.
>
> Your graphene example is trivial to solve using a new message type
> "graphene-capabilities"
> With a "minimum version" and a "maximum supported version" int-pair and
> maybe other booleans that your algorithms can use.
> You only send this if the service bit was set, there is generally no reply
> needed as the other side does the same in reply to the VERSION message.
> Probably takes about half an hour to implement.
>
> If you think, Oh, I have to update protocol.c/h and main.cpp and maybe
> other files just to support one new message, then I think you should
> consider refactoring this stuff and lower the complexity on the BU codebase
> instead of increasing the complexity for everyone by introducing a new
> layer of abstraction on the network.
>
> It should also be mentioned that to me it looks like BU isn't releasing
> more than one Graphene version per 6 months to the wider public. Which
> means that you have no requirement for heavy versioning in the first place
> since there are hard forks every 6 months. So this problem is one of
> planning when to do this upgrade, not any real versioning problem.
>
> Hope that helps.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1236#issuecomment-485568424>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAV662PE72SKTTFKBIN3NTPRYY5TANCNFSM4FNMCOYA>
> .
>

zander commented:
Hi Andrew,

> Regardless, XVERSION has been in BU since the last release so this ship has sailed.

Ok, just for clarity, just this week my opinion was asked on this draft new message type. I had not seen this before a couple of days ago. Either the draft, not any hint of this work has been seen on any of my normal channels, there has been no personal outreach.

This xversion message, from your documentation, is meant to be used by ALL clients. And I was indeed asked directly if I want to implement it in Flowee the Hub.

And now you tell me that you can handle no feedback or even improvement suggestions because **"that ship has sailed"** ?

I will just close here by saying that your design goals of making this cross-client is completely at odds with your behaviour. This interaction did not make me feel good. This is not a way of working that I like to continue.

To conclude:

As far as you have been able to show, xversion does not solve any problems, it just shuffles them a bit to a higher abstraction layer by adding complexity, with no real benefits you have shown or I can find.


ps: I had a chuckle at this logic;

> Sending a new message type will get you banned by some clients.

and then later in the same reply

> You must be aware that we have added new messages to the protocol since XVERSION itself is one.

Greg-Griffith commented:
@zander Constructive feedback is always helpful for us so I'd like to thank you for sharing your thoughts on xversion. I only have one more question for you. 

We currently use xversion to communicate graphene information to our peers as part of the initial connection handshake. Assuming you were looking to implement graphene into your client, would xversion deter you from doing so if it was required as a prerequisite?

zander commented:
If it is required to do a handshake in order to use graphene my personal wish would value a simple, new message-type to communicate that.

gandrewstone commented:
Tom, at the top you can see that we solicited input from ABC and XT devs.
I believe that we also described this on telegram and in the interop
meetings (yes its quite old).  I think that we likely didn't include you
because you had stopped developing bitcoin classic.  Personally, I was not
aware for quite some time that flowee was effectively a continuation/rename
of classic (a full node), so I could guess that maybe awemany (the original
author of this) was similarly unaware.

WRT your chuckle about adding a message.  Yes, we broke these clients one
last time, and perhaps we should have created a service bit that indicates
XVERSION compatibility.  But honestly I'm not sure if anyone in this
community would have respected our bit since it seems entirely dominated by
might-makes-right thinking.  With the existence of the XVERSION message
nobody needs to respect eachother's bit choices again.

I'm sorry that you feel that this message is redundant.  In the messages
above we have explained why it is better than the prior system.  You have
not debunked this.  To me, your argument seems to be that these advantages
are not necessary.  Ok, you are entitled to your opinion but at the same
time I am entitled to mine.  I feel that this message will be extremely
valuable for the permissionless deployment of new services.  Since this
message is only used for new and optional services you may choose to not
support it or the services that use it.

But the complexity of this message is not much greater than a custom
graphene version handshake message.  Yes, we are sending a vector of key
value pairs, instead of just values.  But vectors are already used in many
places.  If you wanted to hard code a message creation that is compliant
with XVERSION yet uses the streaming (<<) operators, you could do with the
same simple code that a new message type would use.  Reading an XVERSION
would be maybe 3 lines of code harder, requiring a loop where you just
ignore any keys that don't interest you.


On Tue, Apr 23, 2019 at 7:18 AM Griffith <notifications@github.com> wrote:

> @zander <https://github.com/zander> Constructive feedback is always
> helpful for us so I'd like to thank you for sharing your thoughts on
> xversion. I only have one more question for you.
>
> We currently use xversion to communicate graphene information to our peers
> as part of the initial connection handshake. Assuming you were looking to
> implement graphene into your client, would xversion deter you from doing so
> if it was required as a prerequisite?
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1236#issuecomment-485761009>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAAV663KYI3YJJ72GKMPJWDPR3V7LANCNFSM4FNMCOYA>
> .
>
